lib/OpenQA/Script/CloneJob.pm

```diff
@@ -90,21 +90,45 @@ sub clone_job_get_job ($jobid, $remote, $remote_url, $options) {
     return $job;
 }
 
+sub _job_setting_is ($job, $key, $expected_value) {
+    my $actual_value = $job->{settings}->{$key};
+    return $actual_value && $actual_value eq $expected_value;
+}
+
+sub _get_chained_parents ($job, $remote, $remote_url, $options, $parents = [], $parent_ids = {}) {
+    next if $parent_ids->{$job->{id}}++;
+    my @direct_parents = map { clone_job_get_job($_, $remote, $remote_url, $options) } @{$job->{parents}->{Chained}};
+    push @$parents, @direct_parents;
+    _get_chained_parents($_, $remote, $remote_url, $options, $parents, $parent_ids) for @direct_parents;
+    return $parents;
+}
+
+sub _is_asset_generated_by_cloned_jobs ($job, $parents, $file, $options) {
+    return 0 if ($options->{'skip-deps'} || $options->{'skip-chained-deps'}) && (scalar @$parents != 0);
+    for my $j (@$parents, $job) {
+        for my $setting (qw(PUBLISH_HDD_1 PUBLISH_PFLASH_VARS)) {
+            return 1 if _job_setting_is $j, $setting => $file;
+        }
+    }
+    return 0;
+}
+
 sub clone_job_download_assets ($jobid, $job, $remote, $remote_url, $ua, $options) {
-    my @parents = map { clone_job_get_job($_, $remote, $remote_url, $options) } @{$job->{parents}->{Chained}};
+    my $parents = _get_chained_parents($job, $remote, $remote_url, $options);
   ASSET:
     for my $type (keys %{$job->{assets}}) {
         next if $type eq 'repo';    # we can't download repos
         for my $file (@{$job->{assets}->{$type}}) {
             my $dst = $file;
-            # skip downloading published assets assuming we are also cloning
-            # the generation job or if the only cloned job *is* the generation
-            # job.
-            my $nr_parents = @parents;
-            if ((!$options->{'skip-deps'} && !$options->{'skip-chained-deps'}) || ($nr_parents == 0)) {
-                for my $j (@parents, $job) {
-                    next ASSET if $j->{settings}->{PUBLISH_HDD_1} && $file eq $j->{settings}->{PUBLISH_HDD_1};
-                }
+            # skip downloading published assets if we are also cloning the generation job or
+            # if the only cloned job *is* the generation job
+            next ASSET if _is_asset_generated_by_cloned_jobs $job, $parents, $file, $options;
+            # skip downloading "uefi-vars" assets if not actually generated by
+            # any parent
+            if ($file =~ qr/uefi-vars/) {
+                my $parent_publishes_pflash_vars;
```

Comment:
why declare the variable here and not immediately assign? If this is just to prevent line-break then maybe you need to extract more functions to reduce indentation further :)

Reply:
I don't understand. To me it looks like you can combine this with the next line as you unconditionally assign a value there, don't you?

Reply:
The next line is a for loop and the lifetime of the variable is outside of that loop. Even if I could get away with it in Perl this seems conceptionally wrong to me.